graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,386 @@
+                    final Object rawPayload = eventMap.get("payload");
+                    if (rawPayload != null) {
+                        connectionCtx = toMapFromJson(rawPayload);
+                    }
+                    connectionInitiated = true;
+                    writeConnectionAck(out);
+                    break;
+                case "ping":
+                    writePong(out);
+                    break;
+                case "pong":
+                    break;
+                case "subscribe":
+                    ensureInitiated();
+                    id = toStringFromJson(eventMap.get("id"));
+                    if (id == null) {
+                        throw new WebSocketCloseException(4400);
+                    }
+                    final Map<String, Object> payload = toMapFromJson(eventMap.get("payload"));
+                    try {
+                        if (graphqlSubscriptions.containsKey(id)) {
+                            // Subscription already exists
+                            throw new WebSocketCloseException(4409);
+                        }
+                        final String operationName = toStringFromJson(payload.get("operationName"));
+                        final String query = toStringFromJson(payload.get("query"));
+                        final Map<String, Object> variables = toMapFromJson(payload.get("variables"));
+                        final Map<String, Object> extensions = toMapFromJson(payload.get("extensions"));
+
+                        final ExecutionInput.Builder executionInput =
+                                ExecutionInput.newExecutionInput()
+                                              .graphQLContext(connectionCtx)
+                                              .graphQLContext(upgradeCtx)
+                                              .query(query)
+                                              .variables(variables)
+                                              .operationName(operationName)
+                                              .extensions(extensions);
+
+                        final CompletableFuture<ExecutionResult> future =
+                                graphqlExecutor.executeGraphql(ctx, executionInput);
+
+                        future.handleAsync((executionResult, throwable) -> {
+                            handleExecutionResult(out, id, executionResult, throwable);
+                            return null;
+                        }, ctx.eventLoop());
+                    } catch (WebSocketCloseException e) {
+                        logger.debug("Error handling subscription", e);
+                        // Also cancel subscription if present before closing websocket
+                        final ExecutionResultSubscriber s = graphqlSubscriptions.remove(id);
+                        if (s != null) {
+                            s.setCompleted();
+                        }
+                        out.close(e.getWebSocketCloseStatus());
```

Comment:
I think we have to cancel all subscriptions instead of just one because we cannot send the message anyway.
So what do you think of removing this catch clause and handle `WebSocketCloseException` at line 179?
We might want to add a method (e.g. `cleanup()`) and call there:

```java
private void cleanup() {
    for (ExecutionResultSubscriber executionResultSubscriber : graphqlSubscriptions.values()) {
        executionResultSubscriber.setCompleted();
    }
}
```

Reply:
`WebSocketCloseException` is not an IOException. Should we send messages after that?